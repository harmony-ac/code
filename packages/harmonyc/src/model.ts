import { Routers } from './Router'
import { indent } from './util/indent'

export abstract class Branch {
  parent?: Branch
  children: Branch[]
  isFork = false
  isEnd = false

  constructor(children: Branch[] = []) {
    this.children = children
    children.forEach((child) => (child.parent = this))
  }
  setFork(isFork: boolean) {
    this.isFork = isFork
    return this
  }
  setFeatureName(featureName: string) {
    for (const child of this.children) child.setFeatureName(featureName)
    return this
  }
  addChild<C extends Branch>(child: C, index = this.children.length) {
    this.children.splice(index, 0, child)
    child.parent = this
    return child
  }
  get isLeaf() {
    return this.children.length === 0
  }
  get successors(): Branch[] {
    if (!this.isLeaf) return this.children.filter((c, i) => i === 0 || c.isFork)
    else {
      if (this.isEnd) return []
      const next = this.nextNonForkAncestorSibling
      if (next) return [next]
      return []
    }
  }
  get nextNonForkAncestorSibling(): Branch | undefined {
    if (!this.parent) return undefined
    const { nextSibling } = this
    if (nextSibling && !nextSibling.isFork) return nextSibling
    return this.parent.nextNonForkAncestorSibling
  }
  get nextSibling(): Branch | undefined {
    if (!this.parent) return undefined
    return this.parent.children[this.siblingIndex + 1]
  }
  get siblingIndex() {
    return this.parent?.children.indexOf(this) ?? -1
  }
}

export class Step extends Branch {
  action: Action
  responses: Response[]
  state?: State

  constructor(
    action = '',
    responses: string[] = [],
    children?: Branch[],
    isFork = false,
    featureName = ''
  ) {
    super(children)
    this.action = new Action(action, featureName)
    this.responses = responses.map(
      (response) => new Response(response, featureName)
    )
    this.isFork = isFork
  }
  get phrases(): Phrase[] {
    return [this.action, ...this.responses]
  }
  toGherkin() {
    return this.phrases.flatMap((phrase) => phrase.toGherkin())
  }
  setFeatureName(featureName: string) {
    this.action.setFeatureName(featureName)
    for (const response of this.responses) response.setFeatureName(featureName)
    return super.setFeatureName(featureName)
  }
}
export class State {
  text: string

  constructor(text = '') {
    this.text = text
  }
}

export class Label {
  text: string
  constructor(text = '') {
    this.text = text
  }
}

export class Section extends Branch {
  label: Label
  constructor(label = '', children?: Branch[], isFork = false) {
    super(children)
    this.label = new Label(label)
    this.isFork = isFork
  }
}

export abstract class Phrase {
  text: string
  featureName: string
  abstract get kind(): string
  constructor(text = '', featureName = '') {
    this.text = text
    this.featureName = featureName
  }
  setFeatureName(featureName: string) {
    this.featureName = featureName
  }
  toGherkin() {
    return [
      `${this.kind === 'action' ? 'When' : 'Then'} ${this.text} || ${
        this.featureName
      }`,
    ]
  }
}

export class Action extends Phrase {
  kind = 'action'
}

export class Response extends Phrase {
  kind = 'response'
}

export class Precondition extends Branch {
  state: State = new State()
  constructor(state: string = '') {
    super()
    this.state.text = state
  }
}

export function makeTests(root: Branch) {
  const routers = new Routers(root)
  const tests = []
  let ic = routers.getIncompleteCount()
  let newIc: number
  do {
    const newTest = new Test(root, routers.nextWalk())
    newIc = routers.getIncompleteCount()
    if (newIc < ic) tests.push(newTest)
    ic = newIc
  } while (ic > 0)
  // sort by order of appearance of the last branch
  const branchIndex = new Map<Branch, number>()
  let i = 0
  function walk(branch: Branch) {
    branchIndex.set(branch, i++)
    for (const child of branch.children) walk(child)
  }
  walk(root)
  tests.sort((a, b) => branchIndex.get(a.last)! - branchIndex.get(b.last)!)

  tests.forEach((test, i) => (test.testNumber = `T${i + 1}`))
  return tests
}

export class Test {
  testNumber?: string
  constructor(public root: Branch, public branches: Branch[]) {}

  get steps(): Step[] {
    return this.branches.filter((b): b is Step => b instanceof Step)
  }

  get last() {
    return this.steps[this.steps.length - 1]
  }

  get labels(): string[] {
    return this.branches
      .filter((b): b is Section => b instanceof Section)
      .map((s) => s.label.text)
  }

  toGherkin() {
    return [
      `Scenario: ${this.testNumber!}${
        this.labels.length > 0 ? ' - ' : ''
      }${this.labels.join(' - ')}`,
      ...indent(this.steps.flatMap((b) => b.toGherkin())),
      '',
    ]
  }
}
