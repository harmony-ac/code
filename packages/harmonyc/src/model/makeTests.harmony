+ no tests:
  + "" => ""

+ basic cases:
  + | - hello
    =>
    | + T1:
    |   - hello

  + | - hello
    | - world
    =>
    | + T1:
    |   - hello
    |   - world

+ fork:
  + | + hello
    | + world
    =>
    | + T1:
    |   - hello
    | + T2:
    |   - world

+ labels:
  + one:
    + | + greeting:
      |   - hi
      =>
      | + T1 - greeting:
      |   - hi
  + more:
    + | + greeting:
      |   + formal:
      |     - good morning
      =>
      | + T1 - greeting - formal:
      |   - good morning
  + nested: