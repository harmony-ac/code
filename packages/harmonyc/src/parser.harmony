+ production "ACTION"
  + | hello
    =>
    | hello

  + | hello "John"
    =>
    | hello "John"
    
  + | | asdf
    =>
    | | asdf

  + | 
    | | asdf
    =>
    | | asdf

+ production "RESPONSE_ITEM"
  + | => hello
    =>
    | hello

  + | => "Jon"
    =>
    | "Jon"

  + | => | asdf
    =>
    | | asdf

  + newline after "=>":
    + | => 
      | | asdf
      =>
      | | asdf

    + | =>
      | | asdf
      =>
      | | asdf

+ production "TEST_DESIGN"
  + | hello
    => | + hello
  + "hello => world"    => | + hello => world
  + "hello\n  => world" => | + hello => world
  + "hello\n  => world" => | + hello => world

  + string:
    + | hello"John"
      =>| + hello "John"

  + multiline string:
    + | | asdf
      => | + | asdf
    + | | asdf
      | | bsdf
      => | + | asdf
         | | bsdf
    + "hello\n| John" => "+ hello\n| John"
    + "hello\n| John\n| Jim" => "+ hello\n| John\n| Jim"
    + "hello\n| John\n    | Jim" => "+ hello\n| John\n| Jim"

    + | hello
      | world
      =>
      | + hello
      | + world

    + | + hello
      | + world
      => | + hello
         | + world

    + | + hello
      | + world
      => | + hello
         | + world

  + structure:
    + sequence:
      + | - hello
        | - world
        =>
        | - hello
        | - world
    + fork:
      + | + hello
        | + world
        =>
        | + hello
        | + world
    + fork and sequence:
      + | + hello
        | - world
        =>
        | + hello
        | - world

      + | - hello
        | + world
        =>
        | - hello
        | + world

    + indent:
      + | + hello
        |   + world
        =>
        | + hello
        |   + world
      
    + label:
      + | + label:
        |   + world
        =>
        | + label:
        |   + world
    
  
