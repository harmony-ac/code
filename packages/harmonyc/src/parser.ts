import type { Token } from 'typescript-parsec'
import {
  alt_sc,
  apply,
  expectEOF,
  expectSingleResult,
  kright,
  opt_sc,
  rep_sc,
  rule,
  seq,
  tok,
  list_sc,
  kleft,
  kmid,
} from 'typescript-parsec'
import { T, lexer } from './lexer'
import type { Branch } from './model'
import {
  Action,
  Response,
  CodeLiteral,
  StringLiteral,
  Section,
  Step,
  Word,
  Label,
} from './model'

export function parse(input: string) {
  const tokens = lexer.parse(input)
  return expectSingleResult(expectEOF(TEST_DESIGN.parse(tokens)))
}

const NEWLINE = tok(T.Newline)
const SPACE = tok(T.Space)
const WORD = apply(tok(T.Word), ({ text }) => new Word(text))
const DOUBLE_QUOTE_STRING = apply(
  tok(T.DoubleQuoteString),
  ({ text }) => new StringLiteral(JSON.parse(text))
)
const BACKTICK_STRING = apply(
  tok(T.BacktickString),
  ({ text }) => new CodeLiteral(text.slice(1, -1))
)
const PART = alt_sc(WORD, DOUBLE_QUOTE_STRING, BACKTICK_STRING)
const PHRASE = list_sc(PART, rep_sc(SPACE))
const ACTION = apply(PHRASE, (parts) => new Action(parts))
const RESPONSE = apply(PHRASE, (parts) => new Response(parts))
const ARROW = kmid(
  rep_sc(alt_sc(SPACE, NEWLINE)),
  tok(T.ResponseArrow),
  rep_sc(SPACE)
)

const STEP = apply(
  seq(ACTION, rep_sc(kright(ARROW, RESPONSE))),
  ([action, responses]) => new Step(action, responses).setFork(true)
)

const LABEL = apply(
  kleft(list_sc(PART, rep_sc(SPACE)), tok(T.Colon)),
  (words) => new Label(words.map((w) => w.toString()).join(' '))
)

const SECTION = apply(LABEL, (text) => new Section(text))
const BRANCH = alt_sc(SECTION, STEP) // section first, to make sure there is no colon after step

const DENTS = apply(
  opt_sc(seq(rep_sc(SPACE), alt_sc(tok(T.Plus), tok(T.Minus)), SPACE)),
  (lineHead) => {
    if (!lineHead) return { dent: 0, isFork: true }
    const [dents, seqOrFork] = lineHead
    return { dent: dents.length / 2, isFork: seqOrFork.kind === T.Plus }
  }
)

const LINE = apply(
  seq(DENTS, BRANCH),
  ([{ dent, isFork }, branch], [start, end]) => ({
    dent,
    branch: branch.setFork(isFork),
  })
)

const TEST_DESIGN = kmid(
  rep_sc(NEWLINE),
  apply(list_sc(LINE, rep_sc(NEWLINE)), (lines) => {
    const startDent = 0
    let dent = startDent
    const root = new Section(new Label(''))
    let parent: Branch = root

    let lineNo = 0
    for (const { dent: d, branch } of lines) {
      ++lineNo
      if (d > dent + 1) {
        throw new Error(`invalid indent ${d} at line ${lineNo}`)
      } else if (d === dent + 1) {
        parent = parent.children[parent.children.length - 1]
        ++dent
      } else if (d < startDent) {
        throw new Error(`invalid indent at line ${lineNo}`)
      } else
        while (d < dent) {
          parent = parent.parent!
          --dent
        }
      parent.addChild(branch)
    }

    return root
  }),
  rep_sc(NEWLINE)
)

function inputText(start: Token<T>, end: Token<T> | undefined) {
  let text = ''
  let t: Token<T> | undefined = start
  while (t && t !== end) {
    text += t.text
    t = t.next
  }
  return text
}
