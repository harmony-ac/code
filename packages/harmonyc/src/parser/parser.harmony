
+ production "ACTION"
  + | hello
    =>
    | hello

  + | hello "John"
    =>
    | hello "John"
    
  + | | asdf
    =>
    | | asdf

  + | 
    | | asdf
    =>
    | | asdf

  + string only:
    + | "hello"
      =>
      | "hello"

  + special characters:
    + | :
      => !! | Unable to consume token: :
    + | a-b
      =>
      | a-b

    + | a - b
      =>
      | a - b

    + | a "" - b
      =>
      | a "" - b

    + | a `3` +b
      =>
      | a `3` +b
      
    + | :b
      =>
      | :b

    + | a+b
      =>
      | a+b

    + | a + b
      =>
      | a + b
       
    + | a: b
      =>
      | a: b

    + | a : b
      =>
      | a : b

    + | a=b
      =>
      | a=b

    + | a = b
      =>
      | a = b

    + | !! x
      =>
      | !! x

    + | !! "error"
      =>
      | !! "error"

+ production "RESPONSE_ITEM"
  + | => hello
    =>
    | hello

  + | => "Jon"
    =>
    | "Jon"

  + | => | asdf
    =>
    | | asdf

  + newline after "=>":
    + | => 
      | | asdf
      =>
      | | asdf

    + | =>
      | | asdf
      =>
      | | asdf

+ production "TEST_DESIGN"
  + | hello
    => | + hello
  + "hello => world"    => | + hello => world
  + "hello\n  => world" => | + hello => world

  + string:
    + | hello"John"
      =>| + hello "John"

  + multiline string:
    + | | asdf
      => | + | asdf
    + | | asdf
      | | bsdf
      => | + | asdf
         | | bsdf
    + "hello\n| John" => "+ hello\n| John"
    + "hello\n| John\n| Jim" => "+ hello\n| John\n| Jim"
    + "hello\n| John\n    | Jim" => "+ hello\n| John\n| Jim"

  + structure:
    + sequence:
      + | - hello
        | - world
        =>
        | - hello
        | - world
    + fork:
      + | + hello
        | + world
        =>
        | + hello
        | + world
    + fork and sequence:
      + | + hello
        | - world
        =>
        | + hello
        | - world

      + | - hello
        | + world
        =>
        | - hello
        | + world

    + indent:
      + | + hello
        |   + world
        =>
        | + hello
        |   + world
      
    + label:
      + | + label:
        |   + world
        =>
        | + label:
        |   + world
      + | + label: 
        |   + world
        =>
        | + label:
        |   + world
    
  
