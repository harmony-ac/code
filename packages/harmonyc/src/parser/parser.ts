import type { Parser, Token } from 'typescript-parsec'
import {
  alt_sc,
  apply,
  expectEOF,
  expectSingleResult,
  kright,
  opt_sc,
  rep_sc,
  rule,
  seq,
  tok,
  list_sc,
  kleft,
  kmid,
  fail,
  nil,
  rep_n,
  alt,
} from 'typescript-parsec'
import { T, lexer } from './lexer.ts'
import type { Branch } from '../model/model.ts'
import {
  Action,
  Response,
  CodeLiteral,
  StringLiteral,
  Section,
  Step,
  Word,
  Label,
  ErrorResponse,
  SaveToVariable,
} from '../model/model.ts'

export function parse(input: string): Section
export function parse<T>(input: string, production: Parser<any, T>): T
export function parse<T>(
  input: string,
  production: Parser<any, T> = TEST_DESIGN as any
) {
  const tokens = lexer.parse(input)
  return expectSingleResult(expectEOF(production.parse(tokens)))
}

export const NEWLINES = list_sc(tok(T.Newline), nil())
export const WORDS = apply(
  tok(T.Words),
  ({ text }) => new Word(text.trimEnd().split(/\s+/).join(' '))
)
export const DOUBLE_QUOTE_STRING = alt_sc(
  apply(
    tok(T.DoubleQuoteString),
    ({ text }) => new StringLiteral(JSON.parse(text))
  ),
  seq(tok(T.UnclosedDoubleQuoteString), fail('unclosed double-quote string'))
) as Parser<T, StringLiteral>
export const BACKTICK_STRING = apply(
  tok(T.BacktickString),
  ({ text }) => new CodeLiteral(text.slice(1, -1))
)
export const DOCSTRING = apply(
  list_sc(tok(T.MultilineString), tok(T.Newline)),
  (lines) => lines.map(({ text }) => text.slice(2)).join('\n')
)
export const ERROR_MARK = tok(T.ErrorMark)
export const VARIABLE = apply(tok(T.Variable), ({ text }) => new Word(text))

export const PART = alt_sc(WORDS, DOUBLE_QUOTE_STRING, BACKTICK_STRING)
export const PHRASE = seq(
  rep_sc(PART),
  opt_sc(kright(opt_sc(NEWLINES), DOCSTRING))
)
export const ACTION = apply(PHRASE, ([parts, docstring]) => {
  return new Action(parts, docstring)
})
export const RESPONSE = apply(PHRASE, ([parts, docstring]) => {
  return new Response(parts, docstring)
})
export const ERROR_RESPONSE = apply(
  seq(ERROR_MARK, PHRASE),
  ([, [parts, docstring]]) => new ErrorResponse(parts, docstring)
)
export const SAVE_TO_VARIABLE = apply(
  VARIABLE,
  (variable) => new SaveToVariable(variable.text.slice(2, -1))
)

export const ARROW = kright(opt_sc(NEWLINES), tok(T.ResponseArrow))

export const RESPONSE_ITEM = kright(
  ARROW,
  alt(RESPONSE, ERROR_RESPONSE, SAVE_TO_VARIABLE)
)
export const STEP = apply(
  seq(ACTION, rep_sc(RESPONSE_ITEM)),
  ([action, responses]) => new Step(action, responses).setFork(true)
)

export const LABEL = apply(
  kleft(list_sc(PART, nil()), tok(T.Colon)),
  (words) => new Label(words.map((w) => w.toString()).join(' '))
)

export const SECTION = apply(LABEL, (text) => new Section(text))
export const BRANCH = alt_sc(SECTION, STEP) // section first, to make sure there is no colon after step

export const DENTS = apply(alt_sc(tok(T.Plus), tok(T.Minus)), (seqOrFork) => {
  return {
    dent: (seqOrFork.text.length - 2) / 2,
    isFork: seqOrFork.kind === T.Plus,
  }
})

export const LINE = apply(
  seq(DENTS, BRANCH),
  ([{ dent, isFork }, branch], [start, end]) => ({
    dent,
    branch: branch.setFork(isFork),
  })
)

export const TEST_DESIGN = kmid(
  rep_sc(NEWLINES),
  apply(
    opt_sc(
      list_sc(
        apply(LINE, (line, [start, end]) => ({ line, start, end })),
        NEWLINES
      )
    ),
    (lines) => {
      const startDent = 0
      let dent = startDent
      const root = new Section(new Label(''))
      let parent: Branch = root

      for (const { line, start } of lines ?? []) {
        const { dent: d, branch } = line
        if (Math.round(d) !== d) {
          throw new Error(
            `invalid odd indent of ${d * 2} at line ${start!.pos.rowBegin}`
          )
        } else if (d > dent + 1) {
          throw new Error(`invalid indent ${d} at line ${start!.pos.rowBegin}`)
        } else if (d === dent + 1) {
          parent = parent.children[parent.children.length - 1]
          ++dent
        } else if (d < startDent) {
          throw new Error(`invalid indent ${d} at line ${start!.pos.rowBegin}`)
        } else
          while (d < dent) {
            parent = parent.parent!
            --dent
          }
        parent.addChild(branch)
      }

      return root
    }
  ),
  rep_sc(NEWLINES)
)

function inputText(start: Token<T>, end: Token<T> | undefined) {
  let text = ''
  let t: Token<T> | undefined = start
  while (t && t !== end) {
    text += t.text
    t = t.next
  }
  return text
}
