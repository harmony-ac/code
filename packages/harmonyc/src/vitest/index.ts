import type { ViteDevServer, Plugin } from 'vite'
import { watchFiles } from '../cli/watch.ts'

export interface HarmonyPluginOptions {
  watchDir: string
}

export default function harmonyPlugin({
  watchDir,
}: HarmonyPluginOptions): Plugin {
  return {
    name: 'harmony',
    configureServer(server) {
      watchFiles([`${watchDir}/**/*.harmony`])
    },
  }
}
