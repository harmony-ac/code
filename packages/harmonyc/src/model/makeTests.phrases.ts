import { parse } from '../parser/parser'
import { makeTests, Test } from './model'
import { expect } from 'vitest'

export default class MakeTestsPhrases {
  async When_(s: string) {
    const root = parse(s)
    return makeTests(root)
  }
  async Then_(s: string, tests: Test[]) {
    expect(tests.join('\n')).toEqual(s)
  }
}
