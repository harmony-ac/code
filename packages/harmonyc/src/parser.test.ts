import { describe, expect, test } from 'vitest'
import { Section, Step } from './model'
import { parse } from './parser'

function check(input: string, expected?: Section) {
  test(input, () => {
    const res = parse(input)
    if (expected) expect(res).toEqual(expected)
    else expect(res).toMatchSnapshot()
  })
}

check('', new Section(''))
check('hello', new Section('', [new Step('hello', [], [], true)]))
check(
  `
> comment
hello
world`.trim(),
  new Section('', [
    new Step('hello', [], [], true),
    new Step('world', [], [], true),
  ])
)
check(
  `
happy:
# comment
- hello
- world`.trim(),
  new Section('', [
    new Section(
      'happy',
      [new Step('hello', [], [], false), new Step('world', [], [], false)],
      true
    ),
  ])
)
check(
  `
happy:
+ hello
+ world`.trim(),
  new Section('', [
    new Section(
      'happy',
      [new Step('hello', [], [], true), new Step('world', [], [], true)],
      true
    ),
  ])
)
check(
  `
happy:
+ hello:
  + world`.trim(),
  new Section('', [
    new Section(
      'happy',
      [new Section('hello', [new Step('world', [], [], true)], true)],
      true
    ),
  ])
)

check(
  `
hello
- wonderful
  - world`.trim(),
  new Section('', [
    new Step(
      'hello',
      [],
      [new Step('wonderful', [], [new Step('world', [], [], false)], false)],
      true
    ),
  ])
)

describe('section', () => {
  check(
    `
log in:
- enter username
- enter password`.trim()
  )
})

describe('response', () => {
  check(
    `
log in:
- enter username and password => success
`.trim()
  )
})
