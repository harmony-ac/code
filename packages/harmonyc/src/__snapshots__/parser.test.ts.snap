// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`response > log in:
- enter username and password => success 1`] = `
Section {
  "children": [
    Section {
      "children": [
        Step {
          "action": Action {
            "kind": "action",
            "text": "enter username and password",
          },
          "children": [],
          "isEnd": false,
          "isFork": false,
          "parent": [Circular],
          "responses": [
            Response {
              "kind": "response",
              "text": "success",
            },
          ],
        },
      ],
      "isEnd": false,
      "isFork": true,
      "label": Label {
        "text": "log in",
      },
      "parent": [Circular],
    },
  ],
  "isEnd": false,
  "isFork": false,
  "label": Label {
    "text": "",
  },
}
`;

exports[`section > log in:
- enter username
- enter password 1`] = `
Section {
  "children": [
    Section {
      "children": [
        Step {
          "action": Action {
            "kind": "action",
            "text": "enter username",
          },
          "children": [],
          "isEnd": false,
          "isFork": false,
          "parent": [Circular],
          "responses": [],
        },
        Step {
          "action": Action {
            "kind": "action",
            "text": "enter password",
          },
          "children": [],
          "isEnd": false,
          "isFork": false,
          "parent": [Circular],
          "responses": [],
        },
      ],
      "isEnd": false,
      "isFork": true,
      "label": Label {
        "text": "log in",
      },
      "parent": [Circular],
    },
  ],
  "isEnd": false,
  "isFork": false,
  "label": Label {
    "text": "",
  },
}
`;
